# Data preparation

This code is to prepare dataset for modelling a classifier to discriminate emotions using MSP dataset: [Link](https://ecs.utdallas.edu/research/researchlabs/msp-lab/MSP-Podcast.html).

The prerequisite of this notebook is to segment the audio files by following the guidance provided in the MSP dataset. `audio_crop.py` file segmented big chunk audio files into 2-4 audio segments and also 8 seconds in average length audio files for train, test, and validation dataset.

This file is just for practice to extract audio features and put all emotion data into csv files (train, test, validation)

In [1]:
import os
import json
import pandas as pd
from audio_crop import file_labels
import soundfile as sf
import opensmile


In [2]:
# save paths
arouse_path = os.path.relpath('./MSP Data/Annotations/Arousal')
dominance_path = os.path.relpath('./MSP Data/Annotations/Dominance')
valence_path = os.path.relpath('./MSP Data/Annotations/Valence')
train_path = os.path.relpath('./Data/train')
validation_path = os.path.relpath('./Data/validation')
test_path = os.path.relpath('./Data/test')
data_path = os.path.relpath('./Data')

In [3]:
train = pd.DataFrame()
test = pd.DataFrame()
validation = pd.DataFrame()

In [4]:
train_file, validation_files, test_files = file_labels()

Number of training labels: 40
Number of validation labels: 14
Number of testing labels: 20


In [5]:
smile = opensmile.Smile(
    feature_set=opensmile.FeatureSet.eGeMAPSv02,
    feature_level=opensmile.FeatureLevel.Functionals,
    num_workers=4
)


In [57]:
file_paths = []
for file in os.listdir(test_path):
    file_paths.append(os.path.join(test_path, file))
validation = smile.process_files(file_paths)

In [58]:
# Errors in segment.json file that there are the same start and end time for MSP-PODCAST_0153 audio
validation[validation.isna().any(axis=1)]
# train

Empty DataFrame
Columns: [F0semitoneFrom27.5Hz_sma3nz_amean, F0semitoneFrom27.5Hz_sma3nz_stddevNorm, F0semitoneFrom27.5Hz_sma3nz_percentile20.0, F0semitoneFrom27.5Hz_sma3nz_percentile50.0, F0semitoneFrom27.5Hz_sma3nz_percentile80.0, F0semitoneFrom27.5Hz_sma3nz_pctlrange0-2, F0semitoneFrom27.5Hz_sma3nz_meanRisingSlope, F0semitoneFrom27.5Hz_sma3nz_stddevRisingSlope, F0semitoneFrom27.5Hz_sma3nz_meanFallingSlope, F0semitoneFrom27.5Hz_sma3nz_stddevFallingSlope, loudness_sma3_amean, loudness_sma3_stddevNorm, loudness_sma3_percentile20.0, loudness_sma3_percentile50.0, loudness_sma3_percentile80.0, loudness_sma3_pctlrange0-2, loudness_sma3_meanRisingSlope, loudness_sma3_stddevRisingSlope, loudness_sma3_meanFallingSlope, loudness_sma3_stddevFallingSlope, spectralFlux_sma3_amean, spectralFlux_sma3_stddevNorm, mfcc1_sma3_amean, mfcc1_sma3_stddevNorm, mfcc2_sma3_amean, mfcc2_sma3_stddevNorm, mfcc3_sma3_amean, mfcc3_sma3_stddevNorm, mfcc4_sma3_amean, mfcc4_sma3_stddevNorm, jitterLocal_sma3nz_amean, jitterLocal_sma3nz_stddevNorm, shimmerLocaldB_sma3nz_amean, shimmerLocaldB_sma3nz_stddevNorm, HNRdBACF_sma3nz_amean, HNRdBACF_sma3nz_stddevNorm, logRelF0-H1-H2_sma3nz_amean, logRelF0-H1-H2_sma3nz_stddevNorm, logRelF0-H1-A3_sma3nz_amean, logRelF0-H1-A3_sma3nz_stddevNorm, F1frequency_sma3nz_amean, F1frequency_sma3nz_stddevNorm, F1bandwidth_sma3nz_amean, F1bandwidth_sma3nz_stddevNorm, F1amplitudeLogRelF0_sma3nz_amean, F1amplitudeLogRelF0_sma3nz_stddevNorm, F2frequency_sma3nz_amean, F2frequency_sma3nz_stddevNorm, F2bandwidth_sma3nz_amean, F2bandwidth_sma3nz_stddevNorm, F2amplitudeLogRelF0_sma3nz_amean, F2amplitudeLogRelF0_sma3nz_stddevNorm, F3frequency_sma3nz_amean, F3frequency_sma3nz_stddevNorm, F3bandwidth_sma3nz_amean, F3bandwidth_sma3nz_stddevNorm, F3amplitudeLogRelF0_sma3nz_amean, F3amplitudeLogRelF0_sma3nz_stddevNorm, alphaRatioV_sma3nz_amean, alphaRatioV_sma3nz_stddevNorm, hammarbergIndexV_sma3nz_amean, hammarbergIndexV_sma3nz_stddevNorm, slopeV0-500_sma3nz_amean, slopeV0-500_sma3nz_stddevNorm, slopeV500-1500_sma3nz_amean, slopeV500-1500_sma3nz_stddevNorm, spectralFluxV_sma3nz_amean, spectralFluxV_sma3nz_stddevNorm, mfcc1V_sma3nz_amean, mfcc1V_sma3nz_stddevNorm, mfcc2V_sma3nz_amean, mfcc2V_sma3nz_stddevNorm, mfcc3V_sma3nz_amean, mfcc3V_sma3nz_stddevNorm, mfcc4V_sma3nz_amean, mfcc4V_sma3nz_stddevNorm, alphaRatioUV_sma3nz_amean, hammarbergIndexUV_sma3nz_amean, slopeUV0-500_sma3nz_amean, slopeUV500-1500_sma3nz_amean, spectralFluxUV_sma3nz_amean, loudnessPeaksPerSec, VoicedSegmentsPerSec, MeanVoicedSegmentLengthSec, StddevVoicedSegmentLengthSec, MeanUnvoicedSegmentLength, StddevUnvoicedSegmentLength, equivalentSoundLevel_dBp]
Index: []

[0 rows x 88 columns]

In [59]:
validation = validation.reset_index(level=[1,2])
validation

start                       end  \
file                                                                   
Data/test/MSP-PODCAST_0381_0331.wav 0 days    0 days 00:00:01.260000   
Data/test/MSP-PODCAST_1183_0025.wav 0 days    0 days 00:00:01.590000   
Data/test/MSP-PODCAST_0381_0334.wav 0 days 0 days 00:00:01.809937500   
Data/test/MSP-PODCAST_0709_0064.wav 0 days    0 days 00:00:04.010000   
Data/test/MSP-PODCAST_0709_0018.wav 0 days 0 days 00:00:16.579937500   
...                                    ...                       ...   
Data/test/MSP-PODCAST_0709_0050.wav 0 days    0 days 00:00:04.060000   
Data/test/MSP-PODCAST_0381_0421.wav 0 days 0 days 00:00:01.879937500   
Data/test/MSP-PODCAST_0190_0024.wav 0 days 0 days 00:00:02.088687500   
Data/test/MSP-PODCAST_0566_0322.wav 0 days 0 days 00:00:07.110062500   
Data/test/MSP-PODCAST_0772_0155.wav 0 days 0 days 00:00:07.690062500   

                                     F0semitoneFrom27.5Hz_sma3nz_amean  \
file                                                                     
Data/test/MSP-PODCAST_0381_0331.wav                          29.733887   
Data/test/MSP-PODCAST_1183_0025.wav                          38.951458   
Data/test/MSP-PODCAST_0381_0334.wav                          23.347149   
Data/test/MSP-PODCAST_0709_0064.wav                          33.340130   
Data/test/MSP-PODCAST_0709_0018.wav                          32.386032   
...                                                                ...   
Data/test/MSP-PODCAST_0709_0050.wav                          35.832962   
Data/test/MSP-PODCAST_0381_0421.wav                          34.128738   
Data/test/MSP-PODCAST_0190_0024.wav                          37.010036   
Data/test/MSP-PODCAST_0566_0322.wav                          39.531269   
Data/test/MSP-PODCAST_0772_0155.wav                          36.323574   

                                     F0semitoneFrom27.5Hz_sma3nz_stddevNorm  \
file                                                                          
Data/test/MSP-PODCAST_0381_0331.wav                                0.175097   
Data/test/MSP-PODCAST_1183_0025.wav                                0.127262   
Data/test/MSP-PODCAST_0381_0334.wav                                0.016608   
Data/test/MSP-PODCAST_0709_0064.wav                                0.074612   
Data/test/MSP-PODCAST_0709_0018.wav                                0.067581   
...                                                                     ...   
Data/test/MSP-PODCAST_0709_0050.wav                                0.066545   
Data/test/MSP-PODCAST_0381_0421.wav                                0.042876   
Data/test/MSP-PODCAST_0190_0024.wav                                0.094824   
Data/test/MSP-PODCAST_0566_0322.wav                                0.082213   
Data/test/MSP-PODCAST_0772_0155.wav                                0.186099   

                                     F0semitoneFrom27.5Hz_sma3nz_percentile20.0  \
file                                                                              
Data/test/MSP-PODCAST_0381_0331.wav                                   24.380022   
Data/test/MSP-PODCAST_1183_0025.wav                                   34.680119   
Data/test/MSP-PODCAST_0381_0334.wav                                   22.966850   
Data/test/MSP-PODCAST_0709_0064.wav                                   30.718527   
Data/test/MSP-PODCAST_0709_0018.wav                                   31.043238   
...                                                                         ...   
Data/test/MSP-PODCAST_0709_0050.wav                                   33.360382   
Data/test/MSP-PODCAST_0381_0421.wav                                   32.994587   
Data/test/MSP-PODCAST_0190_0024.wav                                   34.288326   
Data/test/MSP-PODCAST_0566_0322.wav                                   36.004913   
Data/test/MSP-PODCAST_0772_0155.wav                                   31.836845   

                             

In [60]:
# train.dropna()
validation = validation.dropna()
validation

start                       end  \
file                                                                   
Data/test/MSP-PODCAST_0381_0331.wav 0 days    0 days 00:00:01.260000   
Data/test/MSP-PODCAST_1183_0025.wav 0 days    0 days 00:00:01.590000   
Data/test/MSP-PODCAST_0381_0334.wav 0 days 0 days 00:00:01.809937500   
Data/test/MSP-PODCAST_0709_0064.wav 0 days    0 days 00:00:04.010000   
Data/test/MSP-PODCAST_0709_0018.wav 0 days 0 days 00:00:16.579937500   
...                                    ...                       ...   
Data/test/MSP-PODCAST_0709_0050.wav 0 days    0 days 00:00:04.060000   
Data/test/MSP-PODCAST_0381_0421.wav 0 days 0 days 00:00:01.879937500   
Data/test/MSP-PODCAST_0190_0024.wav 0 days 0 days 00:00:02.088687500   
Data/test/MSP-PODCAST_0566_0322.wav 0 days 0 days 00:00:07.110062500   
Data/test/MSP-PODCAST_0772_0155.wav 0 days 0 days 00:00:07.690062500   

                                     F0semitoneFrom27.5Hz_sma3nz_amean  \
file                                                                     
Data/test/MSP-PODCAST_0381_0331.wav                          29.733887   
Data/test/MSP-PODCAST_1183_0025.wav                          38.951458   
Data/test/MSP-PODCAST_0381_0334.wav                          23.347149   
Data/test/MSP-PODCAST_0709_0064.wav                          33.340130   
Data/test/MSP-PODCAST_0709_0018.wav                          32.386032   
...                                                                ...   
Data/test/MSP-PODCAST_0709_0050.wav                          35.832962   
Data/test/MSP-PODCAST_0381_0421.wav                          34.128738   
Data/test/MSP-PODCAST_0190_0024.wav                          37.010036   
Data/test/MSP-PODCAST_0566_0322.wav                          39.531269   
Data/test/MSP-PODCAST_0772_0155.wav                          36.323574   

                                     F0semitoneFrom27.5Hz_sma3nz_stddevNorm  \
file                                                                          
Data/test/MSP-PODCAST_0381_0331.wav                                0.175097   
Data/test/MSP-PODCAST_1183_0025.wav                                0.127262   
Data/test/MSP-PODCAST_0381_0334.wav                                0.016608   
Data/test/MSP-PODCAST_0709_0064.wav                                0.074612   
Data/test/MSP-PODCAST_0709_0018.wav                                0.067581   
...                                                                     ...   
Data/test/MSP-PODCAST_0709_0050.wav                                0.066545   
Data/test/MSP-PODCAST_0381_0421.wav                                0.042876   
Data/test/MSP-PODCAST_0190_0024.wav                                0.094824   
Data/test/MSP-PODCAST_0566_0322.wav                                0.082213   
Data/test/MSP-PODCAST_0772_0155.wav                                0.186099   

                                     F0semitoneFrom27.5Hz_sma3nz_percentile20.0  \
file                                                                              
Data/test/MSP-PODCAST_0381_0331.wav                                   24.380022   
Data/test/MSP-PODCAST_1183_0025.wav                                   34.680119   
Data/test/MSP-PODCAST_0381_0334.wav                                   22.966850   
Data/test/MSP-PODCAST_0709_0064.wav                                   30.718527   
Data/test/MSP-PODCAST_0709_0018.wav                                   31.043238   
...                                                                         ...   
Data/test/MSP-PODCAST_0709_0050.wav                                   33.360382   
Data/test/MSP-PODCAST_0381_0421.wav                                   32.994587   
Data/test/MSP-PODCAST_0190_0024.wav                                   34.288326   
Data/test/MSP-PODCAST_0566_0322.wav                                   36.004913   
Data/test/MSP-PODCAST_0772_0155.wav                                   31.836845   

                             

In [12]:
file_paths = []
for file in os.listdir(validation_path):
    file_paths.append(os.path.join(validation_path, file))
validation = smile.process_files(file_paths)
validation[validation.isna().any(axis=1)]

/home/changhyun/workspace/ABI_research/env/lib/python3.8/site-packages/opensmile/core/smile.py:308: UserWarning: Segment too short, filling with NaN.
  warnings.warn(


F0semitoneFrom27.5Hz_sma3nz_amean  \
file                                      start  end                                                         
Data/validation/MSP-PODCAST_1188_0020.wav 0 days 0 days 00:00:00.044000                                NaN   

                                                                         F0semitoneFrom27.5Hz_sma3nz_stddevNorm  \
file                                      start  end                                                              
Data/validation/MSP-PODCAST_1188_0020.wav 0 days 0 days 00:00:00.044000                                     NaN   

                                                                         F0semitoneFrom27.5Hz_sma3nz_percentile20.0  \
file                                      start  end                                                                  
Data/validation/MSP-PODCAST_1188_0020.wav 0 days 0 days 00:00:00.044000                                         NaN   

                                                                         F0semitoneFrom27.5Hz_sma3nz_percentile50.0  \
file                                      start  end                                                                  
Data/validation/MSP-PODCAST_1188_0020.wav 0 days 0 days 00:00:00.044000                                         NaN   

                                                                         F0semitoneFrom27.5Hz_sma3nz_percentile80.0  \
file                                      start  end                                                                  
Data/validation/MSP-PODCAST_1188_0020.wav 0 days 0 days 00:00:00.044000                                         NaN   

                                                                         F0semitoneFrom27.5Hz_sma3nz_pctlrange0-2  \
file                                      start  end                                                                
Data/validation/MSP-PODCAST_1188_0020.wav 0 days 0 days 00:00:00.044000                                       NaN   

                                                                         F0semitoneFrom27.5Hz_sma3nz_meanRisingSlope  \
file                                      start  end                                                                   
Data/validation/MSP-PODCAST_1188_0020.wav 0 days 0 days 00:00:00.044000                                          NaN   

                                                                         F0semitoneFrom27.5Hz_sma3nz_stddevRisingSlope  \
file                                      start  end                                                                     
Data/validation/MSP-PODCAST_1188_0020.wav 0 days 0 days 00:00:00.044000                                            NaN   

                                                                         F0semitoneFrom27.5Hz_sma3nz_meanFallingSlope  \
file                                      start  end                                                                    
Data/validation/MSP-PODCAST_1188_0020.wav 0 days 0 days 00:00:00.044000                                           NaN   

                                                                         F0semitoneFrom27.5Hz_sma3nz_stddevFallingSlope  \
file                                      start  end                                                                      
Data/validation/MSP-PODCAST_1188_0020.wav 0 days 0 days 00:00:00.044000                                             NaN   

                                                                         ...  \
file                                      start  end                     ...   
Data/validation/MSP-PODCAST_1188_0020.wav 0 days 0 days 00:00:00.044000  ...   

                                                                         slopeUV0-500_sma3nz_amean  \
file                                      start  end                                                 
Data/validation/MSP-PODCAST_1188_0020.wav 0 days 0 days 00:00:0

In [63]:
indices = []
for index in validation.index:
    indices.append(index[10:-4])

validation['podcast'] = indices
validation = validation.set_index('podcast')
validation

start                       end  \
podcast                                                  
MSP-PODCAST_0381_0331 0 days    0 days 00:00:01.260000   
MSP-PODCAST_1183_0025 0 days    0 days 00:00:01.590000   
MSP-PODCAST_0381_0334 0 days 0 days 00:00:01.809937500   
MSP-PODCAST_0709_0064 0 days    0 days 00:00:04.010000   
MSP-PODCAST_0709_0018 0 days 0 days 00:00:16.579937500   
...                      ...                       ...   
MSP-PODCAST_0709_0050 0 days    0 days 00:00:04.060000   
MSP-PODCAST_0381_0421 0 days 0 days 00:00:01.879937500   
MSP-PODCAST_0190_0024 0 days 0 days 00:00:02.088687500   
MSP-PODCAST_0566_0322 0 days 0 days 00:00:07.110062500   
MSP-PODCAST_0772_0155 0 days 0 days 00:00:07.690062500   

                       F0semitoneFrom27.5Hz_sma3nz_amean  \
podcast                                                    
MSP-PODCAST_0381_0331                          29.733887   
MSP-PODCAST_1183_0025                          38.951458   
MSP-PODCAST_0381_0334                          23.347149   
MSP-PODCAST_0709_0064                          33.340130   
MSP-PODCAST_0709_0018                          32.386032   
...                                                  ...   
MSP-PODCAST_0709_0050                          35.832962   
MSP-PODCAST_0381_0421                          34.128738   
MSP-PODCAST_0190_0024                          37.010036   
MSP-PODCAST_0566_0322                          39.531269   
MSP-PODCAST_0772_0155                          36.323574   

                       F0semitoneFrom27.5Hz_sma3nz_stddevNorm  \
podcast                                                         
MSP-PODCAST_0381_0331                                0.175097   
MSP-PODCAST_1183_0025                                0.127262   
MSP-PODCAST_0381_0334                                0.016608   
MSP-PODCAST_0709_0064                                0.074612   
MSP-PODCAST_0709_0018                                0.067581   
...                                                       ...   
MSP-PODCAST_0709_0050                                0.066545   
MSP-PODCAST_0381_0421                                0.042876   
MSP-PODCAST_0190_0024                                0.094824   
MSP-PODCAST_0566_0322                                0.082213   
MSP-PODCAST_0772_0155                                0.186099   

                       F0semitoneFrom27.5Hz_sma3nz_percentile20.0  \
podcast                                                             
MSP-PODCAST_0381_0331                                   24.380022   
MSP-PODCAST_1183_0025                                   34.680119   
MSP-PODCAST_0381_0334                                   22.966850   
MSP-PODCAST_0709_0064                                   30.718527   
MSP-PODCAST_0709_0018                                   31.043238   
...                                                           ...   
MSP-PODCAST_0709_0050                                   33.360382   
MSP-PODCAST_0381_0421                                   32.994587   
MSP-PODCAST_0190_0024                                   34.288326   
MSP-PODCAST_0566_0322                                   36.004913   
MSP-PODCAST_0772_0155                                   31.836845   

                       F0semitoneFrom27.5Hz_sma3nz_percentile50.0  \
podcast                                                             
MSP-PODCAST_0381_0331                                   30.684464   
MSP-PODCAST_1183_0025                                   38.154648   
MSP-PODCAST_0381_0334                                   23.408285   
MSP-PODCAST_0709_0064                                   33.245762   
MSP-PODCAST_0709_0018                                   32.294773   
...                                                           ...   
MSP-PODCAST_0709_0050                                   35.913128   
MSP-PODCAST_0381_0421                                   34.051029   
MSP-PODCAST_0190_0024                                

In [64]:
li = os.listdir(arouse_path)
list1 = []
for file in li:
    if file[0:21] in test_files:
        list1.append(file)
list1.sort()
list1
# li[0][0:21] in test_files


['MSP-Conversation_0047_1_001.csv',
 'MSP-Conversation_0047_1_002.csv',
 'MSP-Conversation_0047_1_003.csv',
 'MSP-Conversation_0047_1_005.csv',
 'MSP-Conversation_0047_1_006.csv',
 'MSP-Conversation_0047_1_007.csv',
 'MSP-Conversation_0047_1_008.csv',
 'MSP-Conversation_0047_2_001.csv',
 'MSP-Conversation_0047_2_003.csv',
 'MSP-Conversation_0047_2_005.csv',
 'MSP-Conversation_0047_2_006.csv',
 'MSP-Conversation_0047_2_007.csv',
 'MSP-Conversation_0047_2_008.csv',
 'MSP-Conversation_0114_1_001.csv',
 'MSP-Conversation_0114_1_002.csv',
 'MSP-Conversation_0114_1_003.csv',
 'MSP-Conversation_0114_1_005.csv',
 'MSP-Conversation_0114_1_006.csv',
 'MSP-Conversation_0114_1_007.csv',
 'MSP-Conversation_0114_1_009.csv',
 'MSP-Conversation_0114_2_001.csv',
 'MSP-Conversation_0114_2_002.csv',
 'MSP-Conversation_0114_2_003.csv',
 'MSP-Conversation_0114_2_005.csv',
 'MSP-Conversation_0114_2_006.csv',
 'MSP-Conversation_0114_2_007.csv',
 'MSP-Conversation_0114_2_009.csv',
 'MSP-Conversation_0114_3_00

In [65]:
arouse_annotation = {}
for elem in list1:
    if elem[0:23] in arouse_annotation.keys():
        arouse_annotation.get(elem[0:23]).append(elem)
    else:
        arouse_annotation[elem[0:23]] = [elem]
arouse_annotation

{'MSP-Conversation_0047_1': ['MSP-Conversation_0047_1_001.csv',
  'MSP-Conversation_0047_1_002.csv',
  'MSP-Conversation_0047_1_003.csv',
  'MSP-Conversation_0047_1_005.csv',
  'MSP-Conversation_0047_1_006.csv',
  'MSP-Conversation_0047_1_007.csv',
  'MSP-Conversation_0047_1_008.csv'],
 'MSP-Conversation_0047_2': ['MSP-Conversation_0047_2_001.csv',
  'MSP-Conversation_0047_2_003.csv',
  'MSP-Conversation_0047_2_005.csv',
  'MSP-Conversation_0047_2_006.csv',
  'MSP-Conversation_0047_2_007.csv',
  'MSP-Conversation_0047_2_008.csv'],
 'MSP-Conversation_0114_1': ['MSP-Conversation_0114_1_001.csv',
  'MSP-Conversation_0114_1_002.csv',
  'MSP-Conversation_0114_1_003.csv',
  'MSP-Conversation_0114_1_005.csv',
  'MSP-Conversation_0114_1_006.csv',
  'MSP-Conversation_0114_1_007.csv',
  'MSP-Conversation_0114_1_009.csv'],
 'MSP-Conversation_0114_2': ['MSP-Conversation_0114_2_001.csv',
  'MSP-Conversation_0114_2_002.csv',
  'MSP-Conversation_0114_2_003.csv',
  'MSP-Conversation_0114_2_005.csv',
 

In [66]:
li2 = os.listdir(valence_path)
list2 = []
for file in li2:
    if file[0:21] in test_files:
        list2.append(file)
list2.sort()
list2

['MSP-Conversation_0047_1_001.csv',
 'MSP-Conversation_0047_1_005.csv',
 'MSP-Conversation_0047_1_006.csv',
 'MSP-Conversation_0047_1_007.csv',
 'MSP-Conversation_0047_1_008.csv',
 'MSP-Conversation_0047_1_009.csv',
 'MSP-Conversation_0047_2_001.csv',
 'MSP-Conversation_0047_2_002.csv',
 'MSP-Conversation_0047_2_003.csv',
 'MSP-Conversation_0047_2_005.csv',
 'MSP-Conversation_0047_2_006.csv',
 'MSP-Conversation_0047_2_007.csv',
 'MSP-Conversation_0047_2_009.csv',
 'MSP-Conversation_0114_1_001.csv',
 'MSP-Conversation_0114_1_005.csv',
 'MSP-Conversation_0114_1_006.csv',
 'MSP-Conversation_0114_1_007.csv',
 'MSP-Conversation_0114_1_008.csv',
 'MSP-Conversation_0114_1_009.csv',
 'MSP-Conversation_0114_2_001.csv',
 'MSP-Conversation_0114_2_005.csv',
 'MSP-Conversation_0114_2_006.csv',
 'MSP-Conversation_0114_2_007.csv',
 'MSP-Conversation_0114_2_008.csv',
 'MSP-Conversation_0114_2_009.csv',
 'MSP-Conversation_0114_2_011.csv',
 'MSP-Conversation_0114_3_001.csv',
 'MSP-Conversation_0114_3_00

In [67]:
li3 = os.listdir(dominance_path)
list3 = []
for file in li3:
    if file[0:21] in test_files:
        list3.append(file)
list3.sort()
list3

['MSP-Conversation_0047_1_001.csv',
 'MSP-Conversation_0047_1_005.csv',
 'MSP-Conversation_0047_1_006.csv',
 'MSP-Conversation_0047_1_007.csv',
 'MSP-Conversation_0047_1_009.csv',
 'MSP-Conversation_0047_1_010.csv',
 'MSP-Conversation_0047_2_001.csv',
 'MSP-Conversation_0047_2_005.csv',
 'MSP-Conversation_0047_2_006.csv',
 'MSP-Conversation_0047_2_007.csv',
 'MSP-Conversation_0047_2_008.csv',
 'MSP-Conversation_0047_2_009.csv',
 'MSP-Conversation_0114_1_001.csv',
 'MSP-Conversation_0114_1_005.csv',
 'MSP-Conversation_0114_1_006.csv',
 'MSP-Conversation_0114_1_007.csv',
 'MSP-Conversation_0114_1_008.csv',
 'MSP-Conversation_0114_1_009.csv',
 'MSP-Conversation_0114_2_001.csv',
 'MSP-Conversation_0114_2_005.csv',
 'MSP-Conversation_0114_2_006.csv',
 'MSP-Conversation_0114_2_007.csv',
 'MSP-Conversation_0114_2_009.csv',
 'MSP-Conversation_0114_2_010.csv',
 'MSP-Conversation_0114_3_001.csv',
 'MSP-Conversation_0114_3_002.csv',
 'MSP-Conversation_0114_3_003.csv',
 'MSP-Conversation_0114_3_00

In [68]:
valence_annotation = {}
dominance_annotation = {}
for elem in list2:
    if elem[0:23] in valence_annotation.keys():
        valence_annotation.get(elem[0:23]).append(elem)
    else:
        valence_annotation[elem[0:23]] = [elem]
for elem in list3:
    if elem[0:23] in dominance_annotation.keys():
        dominance_annotation.get(elem[0:23]).append(elem)
    else:
        dominance_annotation[elem[0:23]] = [elem]
valence_annotation

{'MSP-Conversation_0047_1': ['MSP-Conversation_0047_1_001.csv',
  'MSP-Conversation_0047_1_005.csv',
  'MSP-Conversation_0047_1_006.csv',
  'MSP-Conversation_0047_1_007.csv',
  'MSP-Conversation_0047_1_008.csv',
  'MSP-Conversation_0047_1_009.csv'],
 'MSP-Conversation_0047_2': ['MSP-Conversation_0047_2_001.csv',
  'MSP-Conversation_0047_2_002.csv',
  'MSP-Conversation_0047_2_003.csv',
  'MSP-Conversation_0047_2_005.csv',
  'MSP-Conversation_0047_2_006.csv',
  'MSP-Conversation_0047_2_007.csv',
  'MSP-Conversation_0047_2_009.csv'],
 'MSP-Conversation_0114_1': ['MSP-Conversation_0114_1_001.csv',
  'MSP-Conversation_0114_1_005.csv',
  'MSP-Conversation_0114_1_006.csv',
  'MSP-Conversation_0114_1_007.csv',
  'MSP-Conversation_0114_1_008.csv',
  'MSP-Conversation_0114_1_009.csv'],
 'MSP-Conversation_0114_2': ['MSP-Conversation_0114_2_001.csv',
  'MSP-Conversation_0114_2_005.csv',
  'MSP-Conversation_0114_2_006.csv',
  'MSP-Conversation_0114_2_007.csv',
  'MSP-Conversation_0114_2_008.csv',
 

In [69]:
dominance_annotation

{'MSP-Conversation_0047_1': ['MSP-Conversation_0047_1_001.csv',
  'MSP-Conversation_0047_1_005.csv',
  'MSP-Conversation_0047_1_006.csv',
  'MSP-Conversation_0047_1_007.csv',
  'MSP-Conversation_0047_1_009.csv',
  'MSP-Conversation_0047_1_010.csv'],
 'MSP-Conversation_0047_2': ['MSP-Conversation_0047_2_001.csv',
  'MSP-Conversation_0047_2_005.csv',
  'MSP-Conversation_0047_2_006.csv',
  'MSP-Conversation_0047_2_007.csv',
  'MSP-Conversation_0047_2_008.csv',
  'MSP-Conversation_0047_2_009.csv'],
 'MSP-Conversation_0114_1': ['MSP-Conversation_0114_1_001.csv',
  'MSP-Conversation_0114_1_005.csv',
  'MSP-Conversation_0114_1_006.csv',
  'MSP-Conversation_0114_1_007.csv',
  'MSP-Conversation_0114_1_008.csv',
  'MSP-Conversation_0114_1_009.csv'],
 'MSP-Conversation_0114_2': ['MSP-Conversation_0114_2_001.csv',
  'MSP-Conversation_0114_2_005.csv',
  'MSP-Conversation_0114_2_006.csv',
  'MSP-Conversation_0114_2_007.csv',
  'MSP-Conversation_0114_2_009.csv',
  'MSP-Conversation_0114_2_010.csv'],


In [70]:
segment_path = os.path.relpath('./MSP Data/Time Labels/segments.json')
f = open(segment_path, 'r')
timing_data = json.load(f)
temp = validation.copy()

In [71]:
def mean(list):
    return sum(list) / len(list)

In [72]:
for key in timing_data:
    conv_part = timing_data[key]['Conversation_Part']
    if key not in temp.index:
        continue
    start = timing_data[key]['Start_Time']
    end = timing_data[key]['End_Time']
    arousal = []
    valence = []
    dominance = []
    a_annotations = arouse_annotation.get(conv_part)
    v_annotations = valence_annotation.get(conv_part)
    d_annotations = dominance_annotation.get(conv_part)
    for anno in a_annotations:
        pp = os.path.join(arouse_path, anno)
        df = pd.read_csv(pp, header=8, names=['time', 'arousal'])
        df = df[(start <= df['time']) & (df['time'] <= end)]
        if df.empty:
            continue
        arousal.append(df['arousal'].mean())
    
    for anno in v_annotations:
        pp = os.path.join(valence_path, anno)
        df = pd.read_csv(pp, header=8, names=['time', 'valence'])
        df = df[(start <= df['time']) & (df['time'] <= end)]
        if df.empty:
            continue
        valence.append(df['valence'].mean())
    for anno in d_annotations:
        pp = os.path.join(dominance_path, anno)
        df = pd.read_csv(pp, header=8, names=['time', 'dominance'])
        if(anno == 'MSP-Conversation_0047_2_001.csv'):
            df = df.reset_index()
            df = df.drop(columns=['dominance'])
            df.columns = ['time', 'dominance']
        df = df[(start <= df['time']) & (df['time'] <= end)]
        if df.empty:
            continue
        
        dominance.append(df['dominance'].mean())
    
    score_arousal = mean(arousal)
    score_valence = mean(valence)
    score_dominance = mean(dominance)
    
    temp.loc[key, 'arousal'] = score_arousal
    temp.loc[key, 'valence'] = score_valence
    temp.loc[key, 'dominance'] = score_dominance
    
temp

start                       end  \
podcast                                                  
MSP-PODCAST_0381_0331 0 days    0 days 00:00:01.260000   
MSP-PODCAST_1183_0025 0 days    0 days 00:00:01.590000   
MSP-PODCAST_0381_0334 0 days 0 days 00:00:01.809937500   
MSP-PODCAST_0709_0064 0 days    0 days 00:00:04.010000   
MSP-PODCAST_0709_0018 0 days 0 days 00:00:16.579937500   
...                      ...                       ...   
MSP-PODCAST_0709_0050 0 days    0 days 00:00:04.060000   
MSP-PODCAST_0381_0421 0 days 0 days 00:00:01.879937500   
MSP-PODCAST_0190_0024 0 days 0 days 00:00:02.088687500   
MSP-PODCAST_0566_0322 0 days 0 days 00:00:07.110062500   
MSP-PODCAST_0772_0155 0 days 0 days 00:00:07.690062500   

                       F0semitoneFrom27.5Hz_sma3nz_amean  \
podcast                                                    
MSP-PODCAST_0381_0331                          29.733887   
MSP-PODCAST_1183_0025                          38.951458   
MSP-PODCAST_0381_0334                          23.347149   
MSP-PODCAST_0709_0064                          33.340130   
MSP-PODCAST_0709_0018                          32.386032   
...                                                  ...   
MSP-PODCAST_0709_0050                          35.832962   
MSP-PODCAST_0381_0421                          34.128738   
MSP-PODCAST_0190_0024                          37.010036   
MSP-PODCAST_0566_0322                          39.531269   
MSP-PODCAST_0772_0155                          36.323574   

                       F0semitoneFrom27.5Hz_sma3nz_stddevNorm  \
podcast                                                         
MSP-PODCAST_0381_0331                                0.175097   
MSP-PODCAST_1183_0025                                0.127262   
MSP-PODCAST_0381_0334                                0.016608   
MSP-PODCAST_0709_0064                                0.074612   
MSP-PODCAST_0709_0018                                0.067581   
...                                                       ...   
MSP-PODCAST_0709_0050                                0.066545   
MSP-PODCAST_0381_0421                                0.042876   
MSP-PODCAST_0190_0024                                0.094824   
MSP-PODCAST_0566_0322                                0.082213   
MSP-PODCAST_0772_0155                                0.186099   

                       F0semitoneFrom27.5Hz_sma3nz_percentile20.0  \
podcast                                                             
MSP-PODCAST_0381_0331                                   24.380022   
MSP-PODCAST_1183_0025                                   34.680119   
MSP-PODCAST_0381_0334                                   22.966850   
MSP-PODCAST_0709_0064                                   30.718527   
MSP-PODCAST_0709_0018                                   31.043238   
...                                                           ...   
MSP-PODCAST_0709_0050                                   33.360382   
MSP-PODCAST_0381_0421                                   32.994587   
MSP-PODCAST_0190_0024                                   34.288326   
MSP-PODCAST_0566_0322                                   36.004913   
MSP-PODCAST_0772_0155                                   31.836845   

                       F0semitoneFrom27.5Hz_sma3nz_percentile50.0  \
podcast                                                             
MSP-PODCAST_0381_0331                                   30.684464   
MSP-PODCAST_1183_0025                                   38.154648   
MSP-PODCAST_0381_0334                                   23.408285   
MSP-PODCAST_0709_0064                                   33.245762   
MSP-PODCAST_0709_0018                                   32.294773   
...                                                           ...   
MSP-PODCAST_0709_0050                                   35.913128   
MSP-PODCAST_0381_0421                                   34.051029   
MSP-PODCAST_0190_0024                                

In [73]:
temp[temp.isnull().any(axis=1)]

Empty DataFrame
Columns: [start, end, F0semitoneFrom27.5Hz_sma3nz_amean, F0semitoneFrom27.5Hz_sma3nz_stddevNorm, F0semitoneFrom27.5Hz_sma3nz_percentile20.0, F0semitoneFrom27.5Hz_sma3nz_percentile50.0, F0semitoneFrom27.5Hz_sma3nz_percentile80.0, F0semitoneFrom27.5Hz_sma3nz_pctlrange0-2, F0semitoneFrom27.5Hz_sma3nz_meanRisingSlope, F0semitoneFrom27.5Hz_sma3nz_stddevRisingSlope, F0semitoneFrom27.5Hz_sma3nz_meanFallingSlope, F0semitoneFrom27.5Hz_sma3nz_stddevFallingSlope, loudness_sma3_amean, loudness_sma3_stddevNorm, loudness_sma3_percentile20.0, loudness_sma3_percentile50.0, loudness_sma3_percentile80.0, loudness_sma3_pctlrange0-2, loudness_sma3_meanRisingSlope, loudness_sma3_stddevRisingSlope, loudness_sma3_meanFallingSlope, loudness_sma3_stddevFallingSlope, spectralFlux_sma3_amean, spectralFlux_sma3_stddevNorm, mfcc1_sma3_amean, mfcc1_sma3_stddevNorm, mfcc2_sma3_amean, mfcc2_sma3_stddevNorm, mfcc3_sma3_amean, mfcc3_sma3_stddevNorm, mfcc4_sma3_amean, mfcc4_sma3_stddevNorm, jitterLocal_sma3nz_amean, jitterLocal_sma3nz_stddevNorm, shimmerLocaldB_sma3nz_amean, shimmerLocaldB_sma3nz_stddevNorm, HNRdBACF_sma3nz_amean, HNRdBACF_sma3nz_stddevNorm, logRelF0-H1-H2_sma3nz_amean, logRelF0-H1-H2_sma3nz_stddevNorm, logRelF0-H1-A3_sma3nz_amean, logRelF0-H1-A3_sma3nz_stddevNorm, F1frequency_sma3nz_amean, F1frequency_sma3nz_stddevNorm, F1bandwidth_sma3nz_amean, F1bandwidth_sma3nz_stddevNorm, F1amplitudeLogRelF0_sma3nz_amean, F1amplitudeLogRelF0_sma3nz_stddevNorm, F2frequency_sma3nz_amean, F2frequency_sma3nz_stddevNorm, F2bandwidth_sma3nz_amean, F2bandwidth_sma3nz_stddevNorm, F2amplitudeLogRelF0_sma3nz_amean, F2amplitudeLogRelF0_sma3nz_stddevNorm, F3frequency_sma3nz_amean, F3frequency_sma3nz_stddevNorm, F3bandwidth_sma3nz_amean, F3bandwidth_sma3nz_stddevNorm, F3amplitudeLogRelF0_sma3nz_amean, F3amplitudeLogRelF0_sma3nz_stddevNorm, alphaRatioV_sma3nz_amean, alphaRatioV_sma3nz_stddevNorm, hammarbergIndexV_sma3nz_amean, hammarbergIndexV_sma3nz_stddevNorm, slopeV0-500_sma3nz_amean, slopeV0-500_sma3nz_stddevNorm, slopeV500-1500_sma3nz_amean, slopeV500-1500_sma3nz_stddevNorm, spectralFluxV_sma3nz_amean, spectralFluxV_sma3nz_stddevNorm, mfcc1V_sma3nz_amean, mfcc1V_sma3nz_stddevNorm, mfcc2V_sma3nz_amean, mfcc2V_sma3nz_stddevNorm, mfcc3V_sma3nz_amean, mfcc3V_sma3nz_stddevNorm, mfcc4V_sma3nz_amean, mfcc4V_sma3nz_stddevNorm, alphaRatioUV_sma3nz_amean, hammarbergIndexUV_sma3nz_amean, slopeUV0-500_sma3nz_amean, slopeUV500-1500_sma3nz_amean, spectralFluxUV_sma3nz_amean, loudnessPeaksPerSec, VoicedSegmentsPerSec, MeanVoicedSegmentLengthSec, StddevVoicedSegmentLengthSec, MeanUnvoicedSegmentLength, StddevUnvoicedSegmentLength, equivalentSoundLevel_dBp, arousal, valence, dominance]
Index: []

[0 rows x 93 columns]

In [74]:
temp = temp.sort_values(by='podcast')
temp

start                       end  \
podcast                                                  
MSP-PODCAST_0047_0213 0 days    0 days 00:00:03.730000   
MSP-PODCAST_0047_0214 0 days 0 days 00:00:03.350937500   
MSP-PODCAST_0047_0215 0 days 0 days 00:00:08.008312500   
MSP-PODCAST_0047_0216 0 days    0 days 00:00:05.557750   
MSP-PODCAST_0047_0217 0 days 0 days 00:00:02.786937500   
...                      ...                       ...   
MSP-PODCAST_1183_0040 0 days 0 days 00:00:51.839968750   
MSP-PODCAST_1183_0041 0 days    0 days 00:00:04.500000   
MSP-PODCAST_1183_0042 0 days 0 days 00:00:11.850031250   
MSP-PODCAST_1183_0043 0 days    0 days 00:00:06.660000   
MSP-PODCAST_1183_0044 0 days    0 days 00:00:02.280000   

                       F0semitoneFrom27.5Hz_sma3nz_amean  \
podcast                                                    
MSP-PODCAST_0047_0213                          29.768476   
MSP-PODCAST_0047_0214                          31.579210   
MSP-PODCAST_0047_0215                          27.062445   
MSP-PODCAST_0047_0216                          27.831499   
MSP-PODCAST_0047_0217                          28.495094   
...                                                  ...   
MSP-PODCAST_1183_0040                          36.602207   
MSP-PODCAST_1183_0041                          40.395649   
MSP-PODCAST_1183_0042                          36.635529   
MSP-PODCAST_1183_0043                          39.190174   
MSP-PODCAST_1183_0044                          37.020596   

                       F0semitoneFrom27.5Hz_sma3nz_stddevNorm  \
podcast                                                         
MSP-PODCAST_0047_0213                                0.077006   
MSP-PODCAST_0047_0214                                0.243525   
MSP-PODCAST_0047_0215                                0.060259   
MSP-PODCAST_0047_0216                                0.057339   
MSP-PODCAST_0047_0217                                0.102368   
...                                                       ...   
MSP-PODCAST_1183_0040                                0.167981   
MSP-PODCAST_1183_0041                                0.137749   
MSP-PODCAST_1183_0042                                0.156615   
MSP-PODCAST_1183_0043                                0.169495   
MSP-PODCAST_1183_0044                                0.119105   

                       F0semitoneFrom27.5Hz_sma3nz_percentile20.0  \
podcast                                                             
MSP-PODCAST_0047_0213                                   27.149118   
MSP-PODCAST_0047_0214                                   28.691582   
MSP-PODCAST_0047_0215                                   25.617311   
MSP-PODCAST_0047_0216                                   26.268845   
MSP-PODCAST_0047_0217                                   26.054155   
...                                                           ...   
MSP-PODCAST_1183_0040                                   32.244251   
MSP-PODCAST_1183_0041                                   35.665443   
MSP-PODCAST_1183_0042                                   32.738571   
MSP-PODCAST_1183_0043                                   34.043365   
MSP-PODCAST_1183_0044                                   32.382317   

                       F0semitoneFrom27.5Hz_sma3nz_percentile50.0  \
podcast                                                             
MSP-PODCAST_0047_0213                                   29.417046   
MSP-PODCAST_0047_0214                                   29.689871   
MSP-PODCAST_0047_0215                                   26.914024   
MSP-PODCAST_0047_0216                                   27.746883   
MSP-PODCAST_0047_0217                                   27.784275   
...                                                           ...   
MSP-PODCAST_1183_0040                                   36.233185   
MSP-PODCAST_1183_0041                                   41.493786   
MSP-PODCAST_1183_0042                                

In [75]:

temp.to_csv('test.csv')

In [115]:
# These values are the total number of each sample
print('Total: ', len(timing_data))
print('Train: ', len(os.listdir(train_path)))
print('Validation: ', len(os.listdir(validation_path)))
print('Test: ', len(os.listdir(test_path)))

Total:  5036
Train:  2783
Validation:  750
Test:  1503


However, there are missing values that openSmile cannot fetch acoustic features from certain audio segments:

Validation dataset: MSP-PODCAST_1188_0020 (Segment name) (Total: 1)
Train dataset: 81 segments in MSP-Conversation_0.153_2.wav were missing (Total: 81) => In segments.json, start time and end time are exactly the same which is 0-length audio segments.

Extraction results:
- Train: 2702
- Validation: 749
- Test: 1503


In [134]:
temp.shape

(2702, 93)

In [78]:
path = os.path.relpath('failed.txt')

with open(path) as f:
    lines = f.read().splitlines()
lines = sorted(lines)
lines

['Data/test/MSP-PODCAST_0047_0279.wav',
 'Data/test/MSP-PODCAST_0114_0281.wav',
 'Data/test/MSP-PODCAST_0114_0291.wav',
 'Data/test/MSP-PODCAST_0167_0128.wav',
 'Data/test/MSP-PODCAST_0167_0152.wav',
 'Data/test/MSP-PODCAST_0167_0205.wav',
 'Data/test/MSP-PODCAST_0190_0025.wav',
 'Data/test/MSP-PODCAST_0190_0081.wav',
 'Data/test/MSP-PODCAST_0202_0018.wav',
 'Data/test/MSP-PODCAST_0218_0190.wav',
 'Data/test/MSP-PODCAST_0281_0022.wav',
 'Data/test/MSP-PODCAST_0281_0028.wav',
 'Data/test/MSP-PODCAST_0281_0033.wav',
 'Data/test/MSP-PODCAST_0281_0034.wav',
 'Data/test/MSP-PODCAST_0295_0463.wav',
 'Data/test/MSP-PODCAST_0381_0276.wav',
 'Data/test/MSP-PODCAST_0381_0278.wav',
 'Data/test/MSP-PODCAST_0381_0280.wav',
 'Data/test/MSP-PODCAST_0381_0283.wav',
 'Data/test/MSP-PODCAST_0381_0285.wav',
 'Data/test/MSP-PODCAST_0381_0287.wav',
 'Data/test/MSP-PODCAST_0381_0288.wav',
 'Data/test/MSP-PODCAST_0381_0290.wav',
 'Data/test/MSP-PODCAST_0381_0292.wav',
 'Data/test/MSP-PODCAST_0381_0294.wav',


In [81]:
f = open("failed_sorted.txt", "a")
for line in lines:
    f.write(line+'\n')
f.close()
